src/python/pants/base/build_root.py

```diff
@@ -51,25 +52,23 @@ def path(self):
     return self._root_dir
 
   @path.setter
-  def path(self, root_dir):
+  def path(self, root_dir: str) -> None:
     """Manually establishes the build root for the current workspace."""
     path = os.path.realpath(root_dir)
     if not os.path.exists(path):
       raise ValueError(f'Build root does not exist: {root_dir}')
     self._root_dir = path
 
-  def reset(self):
+  def reset(self) -> None:
     """Clears the last calculated build root for the current workspace."""
     self._root_dir = None
 
-  def __str__(self):
+  def __str__(self) -> str:
     return f'BuildRoot({self._root_dir})'
 
   @contextmanager
-  def temporary(self, path):
+  def temporary(self, path: str) -> Iterator[None]:
```

Comment:
This is not correct:
```python
$ python
Python 3.7.3 (default, May  3 2019, 14:50:05)
[Clang 8.0.2 (https://android.googlesource.com/toolchain/clang 40173bab62ec7462 on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from contextlib import contextmanager
>>> @contextmanager
... def bob():
...   yield
...
>>> a = bob()
>>> type(a)
<class 'contextlib._GeneratorContextManager'>
>>> help(a)

>>> a.next()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: '_GeneratorContextManager' object has no attribute 'next'
>>> dir(a)
['__abstractmethods__', '__call__', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__enter__', '__eq__', '__exit__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', '__weakref__', '_abc_impl', '_recreate_cm', 'args', 'func', 'gen', 'kwds']
>>> next(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '_GeneratorContextManager' object is not an iterator
>>>
```

Probably `typing.ContextManager[None]`? https://docs.python.org/3/library/typing.html#typing.ContextManager

Reply:
See https://stackoverflow.com/a/49736916. The return type of a function decorated with `@contextmanager` is either `Generic` or `Iterable`.

`typing.ContextManager` refers to the protocol for the context manager itself: https://mypy.readthedocs.io/en/latest/protocols.html#context-manager-protocols

Reply:
I don't care about the stackoverflow claim since the repl proves its wrong! What am I missing / how is my repl journey misrepresentative?

Reply:
Ok, my repl is misrepresentative because the typing here is being applied to the wrapped function yet the the repl test applies to the wrapper. This, it seems to me, is a wart of decorators not enclosing their argument like a normal function call, ie:
```
@contextmanager(
  def thing() -> Iterator[None]:
    yield
)
```

This fake syntax seems to make more sense of the wrapped function typing, it clearly belongs to the function passed to the wrapper function and not the result of the wrapper function call.

I'm sure I'll learn to read type ascriptions of wrapped functions correctly eventually but it sure feels un-natural!

Reply:
> The return type of a function decorated with @contextmanager is either Generic or Iterable.

A better way to phrase this to my ears:
```
A @contextmanager can only decorate functions whose return type is either Generic or Iterable.
```

Reply:
> A better way to phrase this to my ears:
>
>> A @contextmanager can only decorate functions whose return type is either Generic or Iterable.

Great description! 

Thank you for investigating this. I agree that it's not ideal nor intuitive.

(As an aside, type hints are seeing active development at least in PEPs to address the other warts! See https://www.python.org/dev/peps/pep-0585/ and https://www.python.org/dev/peps/pep-0604/)